### PR TITLE
Document some of the basic cmake stuff that is unique to us

### PR DESCRIPTION
**Pull request by CMurtagh-LGTM**
Some of the stuff we do is beyond the default cmake, should be documented.

https://deploy-preview-244--nubook.netlify.app/system/foundations/build-system#dependency-declaration

### CODE REVIEW

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 1]
```diff
@@ -566,3 +566,61 @@ target = {
     "asm_object": "elf64",
 }
 ```
+
+### Dependency Declaration
+
+CMake needs to know what dependencies we have and where to find them.
+In a module's folder the `CMakeLists.txt` file tells CMake what is needed to build a module.
+The `nuclear_modlue()` function tells CMake that this directory contains a module that may be added to a role.
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:574]:
```suggestion
The `nuclear_module()` function tells CMake that this directory contains a module that may be added to a role.
```

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 2]
```diff
@@ -566,3 +566,61 @@ target = {
...
+This function may take named parameters.
+
+| Parameter  | Usage                                                                          |
+| ---------- | ------------------------------------------------------------------------------ |
+| LANGUAGE   | The Programming Language that this module is in (currently only c++)           |
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:579]:
```suggestion
| LANGUAGE   | The programming language that this module is in (currently only C++)           |
```

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 3]
```diff
@@ -566,3 +566,61 @@ target = {
...
+| LIBRARIES  | The libraries that this module depends on and must be linked with              |
+| SOURCES    | Extra code outside the `src` directory that must be built                      |
+| DATA_FILES | Extra data files outside the `data` directory that is necessary for the module |
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:582]:
```suggestion
| DATA_FILES | Extra data files outside the `data` directory that the module depends on |
```
Not sure if this is right but the sentence felt incomplete/grammar dodgy

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 4]
```diff
@@ -566,3 +566,61 @@ target = {
...
+
+Now for libraries in particular CMake needs to know where the libraries are located,
+the function `find_package()` is used to tell CMake to find the library.
+
+CMake then looks in the `cmake/Modules/` directory for files like `FindX.cmake` for the logic to find the files.
+CMake can also look for build instruction files in `lib/cmake/X/` like `XConfig.cmake` where `X` is the name of the
+library.
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:589]:
```suggestion
CMake needs to know where libraries are located. The `cmake/Modules/` directory contains files called `FindX.cmake`, where `X` is the name of a library. These files contain the logic for finding library files. The function  `find_package(X)` is used to tell CMake to look for one of these `.cmake` files and run it.

CMake can also look for build instruction files in `/lib/cmake/X/XConfig.cmake` where `X` is the name of the library.
```

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 5]
```diff
@@ -566,3 +566,61 @@ target = {
...
+An example file looks like
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:590]:
```suggestion
An example of one of these files, `FindAravis.cmake` which finds the Aravis library, is shown below
```

File: src/book/02-system/01-foundations/02-build-system.mdx [hunk 6]
```diff
@@ -566,3 +566,61 @@ target = {
...
+| LIBRARIES                | A list instead of a single LIBRARY                                                            |
+| PATH_SUFFIX              | Add subdirectories to look for includes and libraries in the normal search directories        |
+| BINARY                   | A binary that can be used to extract the version number                                       |
+| VERSION_BINARY_ARGUMENTS | The arguments to pass to the binary that will result in a version number being printed        |
+| VERSION_FILE             | A file that contains information on the version number of the library                         |
+| VERSION_REGEX            | The regex used on the version file to extract the version number                              |
+| LINK_TYPE                | The type of linkage that the library will use: `UNKNOWN` is the default and will pick either, |
+|                          | `SHARED` will use dynamic linkage and look for a `.so` library file                           |
+|                          | `STATIC` will use static linkage and look for a `.a` library file                             |
```
Comment (ysims) [inline src/book/02-system/01-foundations/02-build-system.mdx:626]:
```suggestion
|      SHARED       | No arguments; use dynamic linkage and look for a `.so` library file                           |
|        STATIC     | No arguments; use static linkage and look for a `.a` library file                             |
```
maybe something like this?

Comment (ysims) [reply src/book/02-system/01-foundations/02-build-system.mdx:626]: will need formatting tho

File: src/book/02-system/01-foundations/02-build-system.mdx (hunk 3 above)
Comment (CMurtagh-LGTM) [reply src/book/02-system/01-foundations/02-build-system.mdx:582]:
```suggestion
| DATA_FILES | Extra data files outside the `data` directory that the module uses |
```

File: src/book/02-system/01-foundations/02-build-system.mdx (hunk 6 above)
Comment (CMurtagh-LGTM) [reply src/book/02-system/01-foundations/02-build-system.mdx:626]: `Unknown`, `Shared` and `Static` are options to link type.

Comment (ysims) [reply src/book/02-system/01-foundations/02-build-system.mdx:623]: Oh I see, in that case put it on the one line? It looks like boolean flags here, I assumed each row of the table was a separate parameter.

**Review by LachlanCourt — DISMISSED** [review]
Assuming the content is right this seems fine to me :+1:

**Review by LachlanCourt — APPROVED** [review]

**Pull request merged**